### ISSUE TITLE
About category_id -1

### ISSUE THREAD

**Issue by jinfagang** [zylo117/Yet-Another-EfficientDet-Pytorch#298]
In coco default category id starts from 1, in most self-defined dataset, category id is from 0.

Does -1 should remove if trained on such dataset?

**Comment by zylo117** [comment]
？？？
Shouldn't you just add 1 to your dataset to adapt to coco labeling?
Because if you label in coco-styled, it's already one-indexed.
If you don't, you are gonna need to convert it to coco styled anyway.

**Comment by jinfagang** [comment]
what am ask is remove minus `-1` the same effect? the dataset selfdefined converted from VOC to coco and indexed from 0.

**Comment by zylo117** [comment]
Yes, it would be the same. But the official coco category id starts from 1. It would be easier to keep it that way.

**Comment by jinfagang** [comment]
Found wried problem. If remove `-1` operation, the result is totally bad, it messed up location even.

**Comment by zylo117** [comment]
did you modify inference code to adapt to your changes?

**Comment by jinfagang** [comment]
No, nothing changed, simple edited the category_id in dataset.py

**Comment by xiaoshuyun** [comment]
> what am ask is remove minus `-1` the same effect? the dataset selfdefined converted from VOC to coco and indexed from 0.

Hi~would you like to show me the coda that voc to coco？ my email is 845865413@qq.com
thanks a lot.

**Comment by jinfagang** [comment]
@xiaoshuyun  You can using alfred: https://github.com/jinfagang/alfred

**Comment by Cli98** [comment]
> did you modify inference code to adapt to your changes?


@zylo117 That's why I ask you that why you modify code in category when you load data, for the same reason. And @jinfagang the conclusion is just to avoid 0-index in coco annotation.

**Comment by zylo117** [comment]
But it should be any 0 index in coco annotation. If 0 means background, just kick it out. If 0 means the first class of certain object, just add an offset to 1.
I can't make the training program to fit every dataset. It has to be the other way around.

**Comment by Cli98** [comment]
> But it shouldn't be any 0 index in coco annotation. If 0 means background, just kick it out. If 0 means the first class of certain object, just add an offset to 1.
> I can't make the training program to fit every dataset. It has to be the other way around.

Yeah you are right.

**Comment by jinfagang** [comment]
@Cli98  You said we have make dataset coco labeling index from 1, otherwise zylo117's code will minus 1, the everything messed up. I just want to ensure where anything else made a difference than here:

![image](https://user-images.githubusercontent.com/21303438/82636392-1b185280-9c35-11ea-86d3-157a0fb0709d.png)

Which I force assgined 0 for single class (even with zylo117's original code, it should be 0 if coco annoation index from 1). Is that same thing if I forced to 0?

**Comment by Cli98** [comment]
> @Cli98 You said we have to make dataset coco labeling index from 1, otherwise zylo117's code will minus 1, then everything messed up. I just want to ensure where anything else made a difference than here:
> 
> ![image](https://user-images.githubusercontent.com/21303438/82636392-1b185280-9c35-11ea-86d3-157a0fb0709d.png)
> 
> Which I force assgined 0 for single class (even with zylo117's original code, it should be 0 if coco annoation index from 1). Is that same thing if I forced to 0?

@jinfagang Yup. That's my meaning. But I'm not quite suggest you do this modification. As you may not able to track the changes. For my recommendation, try to visualize your detection result to make sure it matches what you consider it should be.

**Comment by jinfagang** [comment]
My problem is that, the training result for single class is work but sometimes miss, don't know is anchor issue or id issue, overall result is not as good as YoloV3 even

**Comment by alexriedel1** [comment]
> Yes, it would be the same. But the official coco category id starts from 1. It would be easier to keep it that way.


Actually COCO Format is 0-indexed (https://cocodataset.org/#format-data) so I don't really get, why it needs to be 1-indexed ?

**Comment by zylo117** [comment]
> > Yes, it would be the same. But the official coco category id starts from 1. It would be easier to keep it that way.
> 
> Actually COCO Format is 0-indexed (https://cocodataset.org/#format-data) so I don't really get, why it needs to be 1-indexed ?

The first class of coco dataset is `person`, whose category id is 1, so it's one-based indexing

**Comment by alexriedel1** [comment]
Sorry your totally right! I'm providing this script to transform 0-indexed cocos to 1-indexed for anyone who's having this issue:
```
from tkinter import Tk
from tkinter import filedialog
import ntpath
import json

Tk().withdraw()

file = filedialog.askopenfilename(title='Choose COCO JSON File ')

with open(file) as f:
    data = json.load(f)
    for cats in data["categories"]:
        cats["id"] = cats["id"] +1
    for anns in data["annotations"]:
        anns["category_id"] = anns["category_id"] + 1

with open(ntpath.dirname(file) + "/" + ntpath.splitext(ntpath.basename(file))[0] + "_1-index.json", "w") as json_file:
    json.dump(data, json_file)
```
